Objects/frameobject.c

```diff
@@ -16,6 +16,622 @@
+        return -1;
+    }
+
+    return 0;
+}
+
+static PyObject*
+framelocalsproxy_contains(PyObject *self, PyObject *key)
+{
+    PyObject* keys = framelocalsproxy_keys(self, NULL);
+
+    if (keys == NULL) {
+        return NULL;
+    }
+
+    int result = PySequence_Contains(keys, key);
+
+    Py_DECREF(keys);
+
+    if (result < 0) {
+        return NULL;
+    }
+
+    return PyBool_FromLong(result);
+}
+
+static PyObject*
+framelocalsproxy_update(PyObject *self, PyObject *other)
+{
+    if (framelocalsproxy_merge(self, other) < 0) {
+        PyErr_SetString(PyExc_TypeError, "update() argument must be dict or another FrameLocalsProxy");
+        return NULL;
+    }
+
+    Py_RETURN_NONE;
+}
+
+static PyObject*
+framelocalsproxy_get(PyObject* self, PyObject *const *args, Py_ssize_t nargs)
+{
+    if (nargs < 1 || nargs > 2) {
+        PyErr_SetString(PyExc_TypeError, "get expected 1 or 2 arguments");
+        return NULL;
+    }
+
+    PyObject *key = args[0];
+    PyObject *default_value = Py_None;
+
+    if (nargs == 2) {
+        default_value = args[1];
+    }
+
+    if (!PyUnicode_CheckExact(key)) {
+        PyErr_SetString(PyExc_TypeError, "key must be a string");
+        return NULL;
+    }
+
+    PyObject *result = framelocalsproxy_getitem(self, args[0]);
+
+    if (result == NULL) {
+        if (PyErr_ExceptionMatches(PyExc_KeyError)) {
+            PyErr_Clear();
+            return Py_XNewRef(default_value);
+        }
+        return NULL;
+    }
+
+    return result;
+}
+
+static PyObject*
+framelocalsproxy_setdefault(PyObject* self, PyObject *const *args, Py_ssize_t nargs)
+{
+    if (nargs < 1 || nargs > 2) {
+        PyErr_SetString(PyExc_TypeError, "setdefault expected 1 or 2 arguments");
+        return NULL;
+    }
+
+    PyObject *key = args[0];
+    PyObject *default_value = Py_None;
+
+    if (nargs == 2) {
+        default_value = args[1];
+    }
+
+    if (!PyUnicode_CheckExact(key)) {
+        PyErr_SetString(PyExc_TypeError, "key must be a string");
+        return NULL;
+    }
+
+    PyObject *result = framelocalsproxy_getitem(self, key);
+
+    if (result == NULL) {
+        if (PyErr_ExceptionMatches(PyExc_KeyError)) {
+            PyErr_Clear();
+            if (framelocalsproxy_setitem(self, key, default_value) < 0) {
+                return NULL;
+            }
+            return Py_XNewRef(default_value);
+        }
+        return NULL;
+    }
+
+    return result;
+}
+
+static PyObject*
+framelocalsproxy_reversed(PyObject *self, PyObject *__unused)
+{
+    PyObject *result = framelocalsproxy_keys(self, NULL);
+    if (PyList_Reverse(result) < 0) {
+        Py_DECREF(result);
+        return NULL;
+    }
+    return result;
+}
+
+static PyObject*
+framelocalsproxy_copy_unsupported(PyObject *self, PyObject *__unused)
+{
+    PyErr_SetString(PyExc_TypeError, "FrameLocalsProxy is uncopyable");
+    return NULL;
+}
+
+static PyObject*
+framelocalsproxy_clear_unsupported(PyObject *self, PyObject *__unused)
```

Comment:
The user would get an message like `FrameLocalsProxy does not have an attribute 'clear'`, instead of a more descriptive message that it should not be "clearable".

Reply:
I'm with Mark here. Usually the Pythonic way is to leave unimplemented functions unimplemented; this allows the user the option to test whether the function is present (using `hasattr`) without trying to call it. There are exceptions, but the `AttributeError` seems clear (:-)) enough.

Reply:
Okay that makes sense. The only outlier is probably `copy`. Without any implementation, that `copy` would raise an error saying `FrameLocalsProxy` does not support pickle. Is that fine or we should be more precise about the fact that it's not copyable?

Reply:
That's what copy says for most uncopyable objects. If we want a better message we should do that in copy.

Reply:
Okay, I can remove all the unsupported functions.